k8s: Disallow unknown fields in CNP & CCNP

christarazi opened the pull request:
See commit msgs.

Reported-by: André Martins <andre@cilium.io>

christarazi committed:
k8s: Disallow unknown fields in CNP & CCNP

Previously, the `x-kubernetes-preserve-unknown-fields` field was set to
true because it was incorrectly thought to have disallowed an empty spec
in a CNP / CCNP (empty rule). In reality, it is not needed at all, and
actually allows additional unknown fields to be permitted. Policies
such as the following would be allowed, bypassing the schema validation
of the CRD (note the `toFQDNs2`).

```
apiVersion: "cilium.io/v2"
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: "denylist"
spec:
  endpointSelector:
    matchLabels:
      k8s-app.guestbook: web
  egress:
    - toEndpoints:
      - matchLabels:
          "k8s:io.kubernetes.pod.namespace": kube-system
          "k8s:k8s-app": kube-dns
      toPorts:
        - ports:
           - port: "53"
             protocol: ANY
          rules:
            dns:
              - matchPattern: "*"
    - toFQDNs2:
        - matchName: "www.google.com"
```

Fixes: 691f831ba8 ("k8s, examples: Preserve unknown fields in {C,CC}NP")

Revert "k8s, examples: Preserve unknown fields in {C,CC}NP"
This reverts commit 691f831ba80467827e49cffb359d151120ccf9a9.

Reported-by: André Martins <andre@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm reviewed (approved):
Please add the Fixes <commit sha> to the commit, after the tests have passed

christarazi committed:
daemon, k8s: Reject empty policies

For context, an empty policy is deemed "empty" if it has no `spec` and
no `specs`. When a policy is applied by the user, it will potentially go
through two levels of validation: K8s apiserver (CRD schema) and agent.
The agent-level validation is only reached if it passes the K8s CRD
validation. Agent-level validation tends to be stricter than K8s CRD
validation.

This commit rejects empty policies at the agent level. An empty policy can
still be admitted by the K8s apiserver, which was not the case since K8s
1.15 and later, due to CRD pruning [1]. Note that empty policies are
only allowed by the apiserver if the policy already exists with the same
name (update operation), because of the previous commit. New policies
that are empty do not pass K8s apiserver validation, and therefore are
rejected (this has always been the case since K8s 1.15).

Previously, empty policies were allowed by the agent. However, they were
allowed / rejected at the K8s apiserver, depending on the K8s version.
Versions 1.14 and earlier admitted empty policies because they did not
have CRD pruning as mentioned in [1]. Versions 1.15 and later rejected
empty policies. (We are referring to policies here, but this applies to
any K8s resource.)

The consequence of this is evident in the next commit which reworks the
end-to-end tests. The summary is again dependant on the K8s version:

 - K8s 1.14 and earlier: empty policies admitted; datapath was configured
   with an "empty" rule, meaning allow all traffic.
 - K8s 1.15 and later: empty policies _not_ admitted; datapath continues
   to be configured by the previous policy.

With this commit, we no longer allow the difference in behavior
described above. This means that regardless of the K8s version, the
datapath will always be configured with the previous policy, upon
"admission" of an empty policy. Essentially, empty policies are no
longer allowed to have an empty rule, **at the agent level**.

However, all empty policies will _still_ be permitted by the K8s
apiserver (as long as it's an update operation). The previous commit is
responsible for that. The reason for that is in Cilium 1.8 and earlier,
we set `x-kubernetes-preserve-unknown-fields` in v1 (or
`preserveUnknownFields` in v1beta1). This forced the apiserver to keep a
policy's (or CR's) empty `spec` or `specs`, which had the consequence of
bypassing the detection of "unknown" fields in the policy (or CR), and
therefore allowed unknown fields. (Unknown fields are fields that we do
not expect to be present in a policy; they can manifest themselves as
user typos.)

Another piece of information to note is that when an empty policy is
applied over an existing policy (same name), then the policy resource on
the K8s side (CNP or CCNP) will be out-of-sync with the actual policy in
the datapath. This is new behavior introduced by this commit.

[1]: https://github.com/kubernetes/kubernetes/pull/77333

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
test: Rework no-specs policy test

Following the previous commit, asserting on whether we are able to apply
the policy is no longer relevant. The previous commit allowed empty
policies to be applied (at the K8s apiserver level), but a parsing error
will be triggered which can be observed in the agent logs. Applying an
empty policy will not overwrite the existing policy, it'll simply
continue enforcing the previous policy. Note, the CNP on the K8s side
will be out-of-sync with the actual policy in the datapath now.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/k8s/apis/cilium.io/v2/types_test.go:
```diff
@@ -348,6 +348,26 @@ func (s *CiliumV2Suite) TestParseSpec(c *C) {
+		},
+	}
+	_, err = empty.Parse()
+	c.Assert(err.Error(), Matches, ".+empty policy")
```
Update this to use the sentinel error (`ErrEmptyCNP`)

aanm commented on pkg/k8s/apis/cilium.io/v2/ccnp_types.go:
```diff
@@ -91,14 +92,14 @@ func (r *CiliumClusterwideNetworkPolicy) Parse() (api.Rules, error) {
 	retRules := api.Rules{}
 
-	if r.Spec != nil {
+	switch {
```
This switch is changing behavior, with this switch a policy with both spec and specs defiled wouldn't return any errors. We should keep this functionality but move the `if r.Spec == nil && r.Specs == nil` to line 94.

christarazi replied:
Good catch, will change.

aanm commented on pkg/k8s/apis/cilium.io/v2/cnp_types.go:
```diff
@@ -234,7 +235,8 @@ func (r *CiliumNetworkPolicy) Parse() (api.Rules, error) {
 
-	if r.Spec != nil {
+	switch {
+	case r.Spec != nil:
```
Same comment left for the CCNP.

aanm commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -265,10 +266,20 @@ func (k *K8sWatcher) updateCiliumNetworkPolicyV2(ciliumNPClient clientset.Interf
+		case ns != "" && !errors.Is(err, cilium_v2.ErrEmptyCNP):
+			fallthrough
+		case ns == "" && !errors.Is(err, cilium_v2.ErrEmptyCCNP):
+			log.WithError(err).WithField(logfields.Object, logfields.Repr(oldRuleCpy)).
+				Warn("Error parsing old CiliumNetworkPolicy rule")
```
to be precise this should be CiliumClusterwideNetworkPolicy :-)
